contrib/database/sql/option.go

```diff
@@ -27,6 +31,82 @@ type config struct {
 	dbmPropagationMode tracer.DBMPropagationMode
 }
 
+func (c *config) checkDBMPropagation(driverName string, driver driver.Driver, dsn string) {
+	if c.dbmPropagationMode == tracer.DBMPropagationModeFull {
+		if dsn == "" {
+			dsn = c.dsn
+		}
+		if dbSystem, ok := dbmFullModeUnsupported(driverName, driver, dsn); ok {
+			log.Warn("Using DBM_PROPAGATION_MODE in 'full' mode is not supported for %s, downgrading to 'service' mode. "+
+				"See https://docs.datadoghq.com/database_monitoring/connect_dbm_and_apm/ for more info.",
+				dbSystem,
+			)
+			c.dbmPropagationMode = tracer.DBMPropagationModeService
+		}
+	}
+}
+
+func dbmFullModeUnsupported(driverName string, driver driver.Driver, dsn string) (string, bool) {
```

Comment:
I like this way more robust implementation. Nice work gathering the packages/prefixes needed to do the detection.